include/boost/math/tools/differential_evolution.hpp

```diff
@@ -0,0 +1,261 @@
+      cost[i] = cost_function(population[i]);
+      if (!isnan(target_value) && cost[i] <= target_value) {
+         target_attained = true;
+      }
+      if (current_minimum_cost && cost[i] < *current_minimum_cost) {
+          *current_minimum_cost = cost[i];
+      }
+      if (queries) {
+        #pragma omp critical
+        queries->push_back(std::make_pair(population[i], cost[i]));
+      }
+    }
+
+    const size_t dimension = bounds_.size();
+    const auto crossover_int = static_cast<decltype(g())>(round(static_cast<double>( (URBG::max)() - (URBG::min)()) * CR_));
+    std::vector<URBG> generators(threads_);
+    for (size_t i = 0; i < threads_; ++i) {
+       generators[i].seed(g());
+    }
+    for (size_t generation = 0; generation < max_generations_; ++generation) {
+      if (cancellation && *cancellation) {
+          break;
+      }
+      if (target_attained) {
+         break;
+      }
+#pragma omp parallel for num_threads(threads_)
+      for (size_t i = 0; i < NP_; ++i) {
+        #if defined(_OPENMP)
+        size_t thread_idx = omp_get_thread_num();
+        #else 
+        size_t thread_idx = 0;
+        #endif
+        auto& gen = generators[thread_idx];
+        size_t r1, r2, r3;
+        do {
+          r1 = gen() % NP_;
+        } while (r1 == i);
+        do {
+          r2 = gen() % NP_;
+        } while (r2 == i || r2 == r1);
+        do {
+          r3 = gen() % NP_;
+        } while (r3 == i || r3 == r2 || r3 == r1);
+        ArgumentContainer trial_vector;
+        if constexpr (has_resize) {
+          trial_vector.resize(bounds_.size());
+        }
+        for (size_t j = 0; j < dimension; ++j) {
+          // See equation (4) of the reference:
+          auto guaranteed_changed_idx = gen() % NP_;
+          if (gen() < crossover_int || j == guaranteed_changed_idx) {
+            auto tmp = population[r1][j] + F_ * (population[r2][j] - population[r3][j]);
+            trial_vector[j] = clamp(tmp, bounds_[j][0], bounds_[j][1]);
+          } else {
+            trial_vector[j] = population[i][j];
+          }
+        }
+        auto trial_cost = cost_function(trial_vector);
+        if (queries) {
+          #pragma omp critical
+          queries->push_back(std::make_pair(trial_vector, trial_cost));
+        }
+ 
+        if (trial_cost < cost[i]) {
+          std::swap(population[i], trial_vector);
+          cost[i] = trial_cost;
+          if (!isnan(target_value) && cost[i] <= target_value) {
+              target_attained = true;
+              // In a single-threaded context, I'd put a break statement here,
+              // but OpenMP does not allow break statements in for loops.
+              // We'll just have to wait until the end of this generation.
+          }
+          if (current_minimum_cost && cost[i] < *current_minimum_cost) {
+              *current_minimum_cost = cost[i];
+          }
+        }
+      }
+    }
+
+    auto it = std::min_element(cost.begin(), cost.end());
```

Comment:
If you're committing to C++20, std::min_element can take an execution policy argument. It should be able to use the same thread pool as OMP if implemented with TBB.